File: c_cpp/union_find.c
// 분리 집합

int N, link[1 << 20], sz[1 << 20];

void reset()
{
	for (int i = 0; i <= N; i++)
	{
		link[i] = i; sz[i] = 1;
	}
}

// 경로 압축

int find(int x)
{
	if (x == link[x])
		return x;

	return link[x] = find(link[x]);
}

// 경로 압축 없음

int find(int x)
{
  while (x != link[x]) x = link[x];
  return x;
}

bool same(int a, int b)
{
	return find(a) == find(b);
}

void united(int a, int b)
{
	a = find(a); b = find(b);
	if (sz[a] < sz[b]) swap(a, b);
	sz[a] += sz[b]; link[b] = a;
}

File: python3/union_find.py
# 분리 집합 알고리즘

def find(x, link):
    to_change = []

    x = link[x]
    while x != link[x]:
        to_change.append(x)
        x = link[x]

    while to_change: link[to_change.pop()] = x

    return x

def united(a, b, link, size):
    a = find(a, link)
    b = find(b, link)

    if a == b: return

    if size[a] < size[b]: a, b = b, a

    size[a] += size[b]
    size[b] = 0
    link[b] = a
    
def reset():
  global link, size
  link = [ i for i in range(N) ]
  size = [ 1 for i in range(N) ]
